mkpasswd.1
.TH MKPASSWD 1 "2019-12-30" "Marco d'Itri" "Debian GNU/Linux"
.SH NAME
mkpasswd \- Overfeatured front end to crypt(3)
.SH SYNOPSIS
.B mkpasswd
.I PASSWORD
.RI [ SALT ]
.SH DESCRIPTION
.B mkpasswd
encrypts the given password with the
.BR crypt (3)
libc function, using the given salt.
.SH OPTIONS
.TP
.BR \-S " STRING\fR, \fP" \-\-salt= \fISTRING\fP
Use the
.I STRING
as salt.
If it begins with
.I $
then it will be passed straight to
.BR crypt (3)
without any checks.
.TP
.BR \-R " NUMBER\fR, \fP" \-\-rounds= \fINUMBER\fP
Use
.I NUMBER
rounds.
This argument is ignored if the method chosen
does not support variable rounds.
For the OpenBSD Blowfish method this is
the logarithm of the number of rounds.
The behavior is undefined if this option is used without
.IR \-\-method .
.TP
.BR \-m " TYPE\fR, \fP" \-\-method= \fITYPE\fP
Compute the password using the
.I TYPE
method.
If
.I TYPE
is
.I help
then the list of available methods is printed.
If
.I TYPE
begins and end with
.I $
characters then the string is passed to
.BR crypt_gensalt (3)
as-is.
.TP
.B \-5
Like
.IR \-\-method=md5crypt .
.TP
.BI \-P " NUM\fR, \fP" \-\-password-fd= NUM
Read the password from file descriptor
.I NUM
instead of using
.BR getpass (3).
If the file descriptor is not connected to a tty then no other text
than the hashed password is printed on stdout.
.TP
.BR \-s ", " \-\-stdin
Like
.IR \-\-password-fd=0 .
.SH ENVIRONMENT
.IP "MKPASSWD_OPTIONS"
A list of options which will be evaluated before the ones specified on the
command line.
.SH BUGS
If the
.I \-\-stdin
option is used then passwords containing some control
characters may not be read correctly.
.P
This program suffers of a bad case of featuritis.
.SH "SEE ALSO"
.BR passwd (1),
.BR passwd (5),
.BR crypt (3),
.BR crypt (5),
.BR crypt_gensalt (3),
.BR getpass (3).
.SH AUTHOR
.B mkpasswd
and this man page were written by Marco d'Itri
.RI < md@linux.it >
and are licensed under the terms of the GNU General Public License,
version 2 or later.
.\" SPDX-License-Identifier: GPL-2.0-or-later


debian/changelog
whois (5.5.23) unstable; urgency=medium

  * Updated the .sc, .新加坡 (.xn--yfro4i67o, Singapore) and .சிங்கப்பூர்
    (.xn--clchc0ea0b2g2a9gcd, Singapore) TLD servers.

 -- Marco d'Itri <md@linux.it>  Sat, 04 May 2024 03:13:22 +0200

whois (5.5.22) unstable; urgency=medium

  * Fixed a segmentation fault with --no-recursion.
  * Updated the .bm and .vi TLD servers.
  * Removed 4 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Mon, 01 Apr 2024 20:14:41 +0200

whois (5.5.21) unstable; urgency=medium

  * Updated the .cv and .sd TLD servers.
  * Remove 4 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Sat, 24 Feb 2024 17:22:22 +0100

whois (5.5.20) unstable; urgency=medium

  * Added the .gn TLD server.
  * Removed 6 new gTLDs which are no longer active.
  * Enabled getopt_long(3) support on Solaris.

 -- Marco d'Itri <md@linux.it>  Fri, 10 Nov 2023 15:26:02 +0100

whois (5.5.19) unstable; urgency=medium

  * Fixed english support for Japanese queries to not add again the /e
    argument if it had already been provided by the user. (Closes: #1050171)
  * Added the .ye and .বাংলা (.xn--54b7fta0cc, Bangladesh) TLD servers.
  * Updated the .ba, .bb, .dk, .es, .gt, .jo, .ml, .mo, .pa, .pn, .sv,
    .uy, .ﺍﻻﺭﺪﻧ (.xn--mgbayh7gpa, Jordan) and .澳門 (.xn--mix891f, Macao)
    TLD servers.
  * Upgraded the TLD URLs to HTTPS whenever possible.
  * Updated the charset for whois.jprs.jp.
  * Removed 3 new gTLDs which are no longer active.
  * Removed support for the obsolete as32 dot notation.

 -- Marco d'Itri <md@linux.it>  Sun, 08 Oct 2023 22:58:33 +0200

whois (5.5.18) unstable; urgency=medium

  [ Marco d'Itri ]
  * Updated the .ga TLD server. (Closes: #1037288)
  * Added new recovered IPv4 allocations.
  * Removed the delegation of 43.0.0.0/8 to JPNIC.
  * Removed 12 new gTLDs which are no longer active.
  * Improved the man page source, courtesy of Bjarni Ingi Gislason.
    (Closes: #1040613)

  [ Robert Scheck ]
  * Added the .edu.za SLD server.
  * Updated the .alt.za SLD server.
  * Added the -ru and -su NIC handles servers.

 -- Marco d'Itri <md@linux.it>  Sat, 22 Jul 2023 17:56:56 +0200

whois (5.5.17) unstable; urgency=medium

  [ Robert Scheck ]
  * Added the .cd TLD server.
  * Updated the -kg NIC handles server name.

  [ Marco d'Itri ]
  * Removed 2 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Wed, 03 May 2023 14:24:37 +0200

whois (5.5.16) unstable; urgency=medium

  * Add bash completion support, courtesy of Ville Skyttä.
  * Updated the .tr TLD server.
  * Removed support for -metu NIC handles.

 -- Marco d'Itri <md@linux.it>  Tue, 28 Feb 2023 00:58:22 +0100

whois (5.5.15) unstable; urgency=medium

  * Updated the .bd, .nz and .tv TLD servers.
  * Added the .llyw.cymru, .gov.scot and .gov.wales SLD servers.
  * Updated the .ac.uk and .gov.uk SLD servers.
  * Recursion has been enabled for whois.nic.tv.
  * Updated the list of new gTLDs with four generic TLDs assigned in
    October 2013 which were missing due to a bug.
  * Removed 4 new gTLDs which are no longer active.
  * Added the Georgian translation, contributed by Temuri Doghonadze.
  * Updated the Finnish translation, contributed by Lauri Nurmi.

 -- Marco d'Itri <md@linux.it>  Wed, 28 Dec 2022 20:44:11 +0100

whois (5.5.14) unstable; urgency=medium

  * Added the .bf and .sd TLD servers.
  * Removed the .gu TLD server.
  * Updated the .dm, .fj, .mt and .pk TLD servers.
  * Updated the charset for whois.nic.tr.
  * Updated the list of new gTLDs.
  * Removed whois.nic.fr from the list of RIPE-like servers, because it
    is not one anymore. (Closes: #1021110)
  * Renamed whois.arnes.si to whois.register.si in the list of RIPE-like
    servers.
  * Added the hiding string for whois.auda.org.au.
  * Fixed uclibc builds, because uclibc does not have NLS support.

 -- Marco d'Itri <md@linux.it>  Sun, 16 Oct 2022 22:16:34 +0200

whois (5.5.13) unstable; urgency=medium

  * Added the .sd TLD server.
  * Updated the list of new gTLDs.
  * Added the Turkish translation, contributed by Oğuz Ersen.

 -- Marco d'Itri <md@linux.it>  Fri, 08 Apr 2022 01:08:55 +0200

whois (5.5.12) unstable; urgency=medium

  * Updated the .pro TLD server, which was totally broken.
  * Fixed the detection of Japanese locales using $LC_MESSAGES.
  * Implemented providing partial salt strings to mkpasswd.
  * Removed 2 new gTLDs which are no longer active.
  * Updated one or more translations. (Closes: #1003597)
  * Enabled full hardening in debian/rules.

 -- Marco d'Itri <md@linux.it>  Wed, 23 Feb 2022 01:03:11 +0100

whois (5.5.11) unstable; urgency=medium

  * Implemented a --no-recursion command line option to disable recursion
    from registrar to registry servers.
  * Updated the .pro, .vu and .xxx TLD servers.
  * Updated the list of new gTLDs.
  * Removed 7 new gTLDs which are no longer active.
  * Updated make_version_h.pl to support Ubuntu no-change uploads,
    contributed by Matthias Klose. (Closes: #995873)

 -- Marco d'Itri <md@linux.it>  Mon, 03 Jan 2022 18:18:36 +0100

whois (5.5.10) unstable; urgency=medium

  * Updated the .lb TLD server.
  * Removed 5 new gTLDs which are no longer active.
  * Updated the charset for whois.lacnic.net, whois.ax, whois.cira.ca
    and whois.dns.pt.
  * Fixed reporting an older version number when using --version.

 -- Marco d'Itri <md@linux.it>  Sun, 06 Jun 2021 19:54:13 +0200

whois (5.5.9) unstable; urgency=medium

  * Updated the .ga TLD server.
  * Removed the .cd and cf TLD servers.
  * Removed 72 new gTLDs which are no longer active.

 -- Marco d'Itri <md@linux.it>  Sun, 28 Mar 2021 00:38:20 +0100

whois (5.5.8) unstable; urgency=medium

  * Added the .xn--4dbrk0ce (.ישראל, Israel) TLD server.
  * Updated the .ie and .az TLD servers.
  * Updated the list of new gTLDs.

 -- Marco d'Itri <md@linux.it>  Tue, 16 Feb 2021 01:53:57 +0100

whois (5.5.7) unstable; urgency=medium

  * Implemented support for CIDR queries to whois.arin.net.

mkpasswd.c
/*
 * Copyright (C) Marco d'Itri <md@linux.it>.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
 */

/* for crypt, snprintf and strcasecmp */
#define _XOPEN_SOURCE 500
#define _BSD_SOURCE 1
#define _DEFAULT_SOURCE 1
#define __EXTENSIONS__ 1

/* System library */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "config.h"
#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#include <fcntl.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <sys/types.h>
#ifdef HAVE_XCRYPT_H
#include <xcrypt.h>
#include <sys/stat.h>
#endif
#ifdef HAVE_CRYPT_H
#include <crypt.h>
#endif
#ifdef HAVE_GETTIMEOFDAY
#include <sys/time.h>
#endif

/* Application-specific */
#include "version.h"
#include "utils.h"

/* Global variables */
#ifdef HAVE_GETOPT_LONG
static const struct option longopts[] = {
    {"method",		optional_argument,	NULL, 'm'},
    /* for backward compatibility with versions < 4.7.25 (< 20080321): */
    {"hash",		optional_argument,	NULL, 'H'},
    {"help",		no_argument,		NULL, 'h'},
    {"password-fd",	required_argument,	NULL, 'P'},
    {"stdin",		no_argument,		NULL, 's'},
    {"salt",		required_argument,	NULL, 'S'},
    {"rounds",		required_argument,	NULL, 'R'},
    {"prompts",		required_argument,	NULL, 'p'},
    {"version",		no_argument,		NULL, 'V'},
    {NULL,		0,			NULL, 0  }
};
#else
extern char *optarg;
extern int optind;
#endif

static const char valid_salts[] = "abcdefghijklmnopqrstuvwxyz"
"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789./";

struct crypt_method {
    const char *method;		/* short name used by the command line option */
    const char *prefix;		/* salt prefix */
    const unsigned int minlen;	/* minimum salt length */
    const unsigned int maxlen;	/* maximum salt length */
    const unsigned int rounds;	/* supports a variable number of rounds */
    const char *desc;		/* long description for the methods list */
};

/* XCRYPT_VERSION_NUM is defined in crypt.h from libxcrypt */
#if defined XCRYPT_VERSION_NUM
# define HAVE_SHA_CRYPT
# define HAVE_BCRYPT
# define HAVE_BSDICRYPT
#endif

static const struct crypt_method methods[] = {
    /* method		prefix	minlen,	maxlen	rounds description */
#ifdef CRYPT_GENSALT_IMPLEMENTS_DEFAULT_PREFIX
    { "auto",		NULL,	0,	0,	0, NULL },
#endif
    /* compatibility aliases for mkpasswd versions < 5.4.0 */
    { "des",		"",	2,	2,	0, NULL },
    { "md5",		"$1$",	8,	8,	0, NULL },
#if defined XCRYPT_VERSION_NUM
    { "yescrypt",	"$y$",	0,	0,	1, "Yescrypt" },
#if XCRYPT_VERSION_NUM >= ((4 << 16) | 4)
    { "gost-yescrypt",	"$gy$",	0,	0,	1, "GOST Yescrypt" },
#endif
    { "scrypt",		"$7$",	0,	0,	1, "scrypt" },
#endif
#ifdef HAVE_BCRYPT_OBSOLETE
    /* http://marc.info/?l=openbsd-misc&m=139320023202696 */
    { "bf",		"$2a$", 22,	22,	2, "bcrypt" },
#endif
#ifdef HAVE_BCRYPT
    { "bcrypt",		"$2b$", 22,	22,	2, "bcrypt" },
    { "bcrypt-a",	"$2a$", 22,	22,	2, "bcrypt (obsolete $2a$ version)" },
#endif
#if defined HAVE_SHA_CRYPT
    /* http://people.redhat.com/drepper/SHA-crypt.txt */
    { "sha512crypt",	"$6$",	8,	16,	1, "SHA-512" },
    { "sha256crypt",	"$5$",	8,	16,	1, "SHA-256" },
    /* compatibility aliases for mkpasswd versions < 5.4.0 */
    { "sha-256",	"$5$",	8,	16,	1, NULL },
    { "sha-512",	"$6$",	8,	16,	1, NULL },
#endif
#if (defined __SVR4 && defined __sun) || defined XCRYPT_VERSION_NUM
    /* http://www.crypticide.com/dropsafe/article/1389 */
    /*
     * Actually the maximum salt length is arbitrary, but Solaris by default
     * always uses 8 characters:
     * http://cvs.opensolaris.org/source/xref/onnv/onnv-gate/ \
     *   usr/src/lib/crypt_modules/sunmd5/sunmd5.c#crypt_gensalt_impl
     */
    { "sunmd5",		"$md5$", 8,	8,	1, "SunMD5" },
#endif
    { "md5crypt",	"$1$",	8,	8,	0, "MD5" },
#ifdef HAVE_BSDICRYPT
    { "bsdicrypt",		"_",	0,	0,	0,
	N_("BSDI extended DES-based crypt(3)") },
#endif
    { "descrypt",	"",	2,	2,	0,
	N_("standard 56 bit DES-based crypt(3)") },
#if defined FreeBSD || defined XCRYPT_VERSION_NUM
    { "nt",		"$3$",  0,	0,	0, "NT-Hash" },
#endif
    { NULL,		NULL,	0,	0,	0, NULL }
};

void generate_salt(char *const buf, const unsigned int len);
void *get_random_bytes(const unsigned int len);
void NORETURN display_help(int error);
void display_version(void);
void display_methods(void);
char *read_line(FILE *fp);

int main(int argc, char *argv[])
{
    int ch, i;
    int password_fd = -1;
    unsigned int salt_minlen = 0;
    unsigned int salt_maxlen = 0;
    unsigned int rounds_support = 0;
    const char *salt_prefix = NULL;
    const char *salt_arg = NULL;
    unsigned int rounds = 0;
    unsigned int prompts = 1;
    char *salt = NULL;
    char rounds_str[30];
    char *password = NULL;

#ifdef ENABLE_NLS
    setlocale(LC_ALL, "");
    bindtextdomain(NLS_CAT_NAME, LOCALEDIR);
    textdomain(NLS_CAT_NAME);
#endif

    /* prepend options from environment */
    argv = merge_args(getenv("MKPASSWD_OPTIONS"), argv, &argc);

    while ((ch = GETOPT_LONGISH(argc, argv, "hH:m:5p:P:R:sS:V", longopts, NULL))
	    > 0) {
	switch (ch) {
	case '5':
	    optarg = (char *) "md5crypt";
	    /* fall through */
	case 'm':
	case 'H':
	    if (!optarg || strcaseeq("help", optarg)) {
		display_methods();
		exit(0);
	    }
#if defined HAVE_LINUX_CRYPT_GENSALT || defined HAVE_SOLARIS_CRYPT_GENSALT
	    if (optarg[0] == '$'
		    && strlen(optarg) > 2
		    && *(optarg + strlen(optarg) - 1) == '$') {
		salt_prefix = NOFAIL(strdup(optarg));
		salt_minlen = 0;
		salt_maxlen = 0;
		rounds_support = 0;
		break;
	    }
#endif
	    for (i = 0; methods[i].method != NULL; i++)
		if (strcaseeq(methods[i].method, optarg)) {
		    salt_prefix = methods[i].prefix;
		    salt_minlen = methods[i].minlen;
		    salt_maxlen = methods[i].maxlen;
		    rounds_support = methods[i].rounds;
		    break;
		}
	    if (!salt_prefix) {
		fprintf(stderr, _("Invalid method '%s'.\n"), optarg);
		exit(1);
	    }
	    break;
	case 'p':
	    {
		char *p;
		long r;

		r = strtol(optarg, &p, 10);
		if (p == NULL || *p != '\0' || r < 0) {
		    fprintf(stderr, _("Invalid number '%s'.\n"), optarg);
		    exit(1);
		}
		prompts = r;
	    }
	    break;
	case 'P':
	    {
		char *p;
		password_fd = strtol(optarg, &p, 10);
		if (p == NULL || *p != '\0' || password_fd < 0) {
		    fprintf(stderr, _("Invalid number '%s'.\n"), optarg);
		    exit(1);
		}
	    }
	    break;
	case 'R':
	    {
		char *p;
		long r;

		r = strtol(optarg, &p, 10);
		if (p == NULL || *p != '\0' || r < 0) {
		    fprintf(stderr, _("Invalid number '%s'.\n"), optarg);
		    exit(1);
		}
		rounds = r;
	    }
	    break;
	case 's':
	    password_fd = 0;
	    break;
	case 'S':
	    salt_arg = optarg;
	    break;
	case 'V':
	    display_version();
	    exit(0);
	case 'h':
	    display_help(EXIT_SUCCESS);
	default:
	    fprintf(stderr, _("Try '%s --help' for more information.\n"),
		    argv[0]);
	    exit(1);
	}
    }
    argc -= optind;
    argv += optind;

    if (argc == 2 && !salt_arg) {
	password = argv[0];
	prompts = prompts > 0 ? prompts - 1 : prompts;
	salt_arg = argv[1];
    } else if (argc == 1) {
	password = argv[0];
	prompts = prompts > 0 ? prompts - 1 : prompts;
    } else if (argc == 0) {
    } else {
	display_help(EXIT_FAILURE);
    }

    /* default: DES password, or else whatever crypt_gensalt chooses */
    if (!salt_prefix) {
	salt_minlen = methods[0].minlen;
	salt_maxlen = methods[0].maxlen;
	salt_prefix = methods[0].prefix;
	rounds_support = methods[0].rounds;
    }

    if (!salt_prefix) {
	/* NULL means that crypt_gensalt will choose one later */
	rounds_str[0] = '\0';
    } else if (rounds_support == 2) {
	/* bcrypt strings always contain the rounds number */
	if (rounds <= 5)
	    rounds = 5;
	/* actually it is the logarithm of the number of rounds */
	snprintf(rounds_str, sizeof(rounds_str), "%02u$", rounds);
    } else if (rounds_support && rounds)
	snprintf(rounds_str, sizeof(rounds_str), "rounds=%u$", rounds);
    else
	rounds_str[0] = '\0';

    if (salt_arg && salt_arg[0] == '$')
	/* the salt begins with the prefix which specifies the method */
	salt = NOFAIL(strdup(salt_arg));
    else if (salt_prefix && salt_arg && strchr(salt_arg, '$')) {
	/* looks like a salt, but with no initial method prefix */
	salt = NOFAIL(malloc(strlen(salt_prefix) + strlen(rounds_str)
		+ strlen(salt_arg) + 1));
	*salt = '\0';
	strcat(salt, salt_prefix);
	strcat(salt, rounds_str);
	strcat(salt, salt_arg);
    } else if (salt_arg && salt_arg[0] != '\0') {
	/* just the salt string with no metadata */
	unsigned int c = strlen(salt_arg);
	if (c < salt_minlen || c > salt_maxlen) {
	    if (salt_minlen == salt_maxlen)
		fprintf(stderr, ngettext(
			"Wrong salt length: %d byte when %d expected.\n",
			"Wrong salt length: %d bytes when %d expected.\n", c),
			c, salt_maxlen);
	    else
		fprintf(stderr, ngettext(
			"Wrong salt length: %d byte when %d <= n <= %d"
			" expected.\n",
			"Wrong salt length: %d bytes when %d <= n <= %d"
			" expected.\n", c),
			c, salt_minlen, salt_maxlen);
	    exit(1);
	}
	while (c-- > 0) {
	    if (strchr(valid_salts, salt_arg[c]) == NULL) {
		fprintf(stderr, _("Illegal salt character '%c'.\n"),
			salt_arg[c]);
		exit(1);
	    }
	}

	/*
	 * Build the actual argument to crypt(3) by concatenating the
	 * method prefix, the rounds metadata (if any) and the salt string.
	 */
	if (!salt_prefix)
	    salt_prefix = "";
	salt = NOFAIL(malloc(strlen(salt_prefix) + strlen(rounds_str)
		+ strlen(salt_arg) + 1));
	*salt = '\0';
	strcat(salt, salt_prefix);
	strcat(salt, rounds_str);
	strcat(salt, salt_arg);
    } else {
	/* no salt was specified by the user, so generate one */
#ifdef HAVE_SOLARIS_CRYPT_GENSALT
	salt = crypt_gensalt(salt_prefix, NULL);
	if (!salt) {
	    perror("crypt_gensalt");
	    exit(2);
	}
#elif defined HAVE_LINUX_CRYPT_GENSALT
	void *entropy = get_random_bytes(64);

	salt = crypt_gensalt(salt_prefix, rounds, entropy, 64);
	if (!salt) {
	    perror("crypt_gensalt");
	    exit(2);
	}
	if (entropy)
	    free(entropy);
#else
	unsigned int salt_len = salt_maxlen;

	if (salt_minlen != salt_maxlen) { /* salt length can vary */
	    srand(time(NULL) + getpid());
	    salt_len = rand() % (salt_maxlen - salt_minlen + 1) + salt_minlen;
	}

	salt = NOFAIL(malloc(strlen(salt_prefix) + strlen(rounds_str)
		+ salt_len + 1));
	*salt = '\0';
	strcat(salt, salt_prefix);
	strcat(salt, rounds_str);
	generate_salt(salt + strlen(salt), salt_len);
#endif
    }

    if (password) {
    } else {
	FILE *fp = NULL;
	if (password_fd != -1) {
	    if (isatty(password_fd))
		fprintf(stderr, _("Password: "));
	    fp = fdopen(password_fd, "r");
	    if (!fp) {
		perror("fdopen");
		exit(2);
	    }
	}

	for (size_t prompt_no = 0; prompt_no < prompts; prompt_no++) {
	    /* getpass will reuse the same buffer, so need to reallocate
	       for multiple prompts */
	    if (!fp && prompt_no == 1) {
		password = NOFAIL(strdup(password));
	    }
	    char * password_n = fp
		? read_line(fp)
		: getpass(_("Password: "));
	    if (!password_n) {
		perror("fgetc");
		exit(2);
	    }
	    if (!password) {
		password = password_n;
	    } else if (!streq(password_n, password)) {
		fprintf(stderr, _("Password did not match.\n"));
		exit(2);
	    }
	}
    }

    {
	const char *result;
	result = crypt(password, salt);
	/* xcrypt returns "*0" on errors */
	if (!result || result[0] == '*') {
	    if (CRYPT_SETS_ERRNO)
		perror("crypt");
	    else
		fprintf(stderr, "crypt failed.\n");
	    exit(2);
	}
	if (!strneq(result, salt, strlen(salt))) {
	    fprintf(stderr, _("Method not supported by crypt(3).\n"));
	    exit(2);
	}
	printf("%s\n", result);
    }

    exit(0);
}

#ifdef CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY

/*
 * If NULL is passed to the libxcrypt version of crypt_gensalt() instead of
 * the buffer of random bytes then the function will obtain by itself the
 * required randomness.
 */
inline void *get_random_bytes(const unsigned int count)
{
    (void)(count); /* suppress the "unused parameter" warning */
    return NULL;
}

#elif defined HAVE_SOLARIS_CRYPT_GENSALT

/*
 * The Solaris version of crypt_gensalt() gathers the random data by itself.
 */

#elif defined RANDOM_DEVICE || defined HAVE_ARC4RANDOM_BUF || defined HAVE_GETENTROPY

void *get_random_bytes(const unsigned int count)
{
    char *buf = NOFAIL(malloc(count));

#if defined HAVE_ARC4RANDOM_BUF
    arc4random_buf(buf, count);
#elif defined HAVE_GETENTROPY
    if (getentropy(buf, count) < 0)
	perror("getentropy");
#else
    int fd;
    ssize_t bytes_read;

    fd = open(RANDOM_DEVICE, O_RDONLY);
    if (fd < 0) {
	perror("open(" RANDOM_DEVICE ")");
	exit(2);
    }
    bytes_read = read(fd, buf, count);
    if (bytes_read < 0) {
	perror("read(" RANDOM_DEVICE ")");
	exit(2);
    }
    if (bytes_read != count) {
	fprintf(stderr, "Short read of %s.\n", RANDOM_DEVICE);
	exit(2);
    }
    close(fd);
#endif

    return buf;
}

void generate_salt(char *const buf, const unsigned int len)
{
    unsigned int i;
    unsigned char *entropy;

    entropy = get_random_bytes(len);

    for (i = 0; i < len; i++)
	buf[i] = valid_salts[entropy[i] % (sizeof valid_salts - 1)];
    buf[i] = '\0';
    free(entropy);
}

#else /* RANDOM_DEVICE || HAVE_ARC4RANDOM_BUF || HAVE_GETENTROPY */

void generate_salt(char *const buf, const unsigned int len)
{
    unsigned int i;

# ifdef HAVE_GETTIMEOFDAY
    struct timeval tv;

    gettimeofday(&tv, NULL);
    srand(tv.tv_sec ^ tv.tv_usec);

# else /* HAVE_GETTIMEOFDAY */
#  warning "This system lacks a strong enough random numbers generator!"

    /*
     * The possible values of time over one year are 31536000, which is
     * two orders of magnitude less than the allowed entropy range (2^32).
     */
    srand(time(NULL) + getpid());

# endif /* HAVE_GETTIMEOFDAY */

    for (i = 0; i < len; i++)
	buf[i] = valid_salts[rand() % (sizeof valid_salts - 1)];
    buf[i] = '\0';
}

#endif /* RANDOM_DEVICE || HAVE_ARC4RANDOM_BUF || HAVE_GETENTROPY*/

void NORETURN display_help(int error)
{
    fprintf((EXIT_SUCCESS == error) ? stdout : stderr,
	    _("Usage: mkpasswd [OPTIONS]... [PASSWORD [SALT]]\n"
	    "Crypts the PASSWORD using crypt(3).\n\n"));
    fprintf((EXIT_SUCCESS == error) ? stdout : stderr, _(
"      -m, --method=TYPE     select method TYPE\n"
"      -5                    like --method=md5crypt\n"
"      -S, --salt=SALT       use the specified SALT\n"
    ));
    fprintf((EXIT_SUCCESS == error) ? stdout : stderr, _(
"      -R, --rounds=NUMBER   use the specified NUMBER of rounds\n"
"      -P, --password-fd=NUM read the password from file descriptor NUM\n"
"                            instead of /dev/tty\n"
"      -s, --stdin           like --password-fd=0\n"
    ));
    fprintf((EXIT_SUCCESS == error) ? stdout : stderr, _(
"      -p  --prompts=NUMBER  prompt NUMBER of times for password.\n"
"                            Useful to guard against typos.\n"
    ));
    fprintf((EXIT_SUCCESS == error) ? stdout : stderr, _(
"      -h, --help            display this help and exit\n"
"      -V, --version         output version information and exit\n"
"\n"
"If PASSWORD is missing then it is asked interactively.\n"
"If no SALT is specified, a random one is generated.\n"
"If TYPE is 'help', available methods are printed.\n"
"\n"
"Report bugs to %s.\n"), "<md+whois@linux.it>");
    exit(error);
}

void display_version(void)
{
    printf("mkpasswd %s\n\n", VERSION);
    puts("Copyright (C) Marco d'Itri\n"
"This is free software; see the source for copying conditions.  There is NO\n"
"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.");
}

void display_methods(void)
{
    unsigned int i;

    printf(_("Available methods:\n"));
    for (i = 0; methods[i].method != NULL; i++)
	if (methods[i].desc)
	    printf("%-15s %s\n", methods[i].method, methods[i].desc);
}

char *read_line(FILE *fp) {
    size_t size = 128;
    int ch;
    size_t pos = 0;
    char *password;

    password = NOFAIL(malloc(size));

    while ((ch = fgetc(fp)) != EOF) {
	if (ch == '\n' || ch == '\r')
	    break;
	password[pos++] = ch;
	if (pos == size) {
	    size += 128;
	    password = NOFAIL(realloc(password, size));
	}
    }
    password[pos] = '\0';

    if (ferror(fp)) {
	free(password);
	return NULL;
    }
    return password;
}

